Python 3 compatibility change to /rmgpy/statmech

xiaoruiDong opened the pull request:
This PR include python 3 compatibility changes to 
```
/rmgpy/statmech/
- __init__.py
- conformerTest.py
- rotationTest.py
- schrodingerTest.py
- torsionTest.py
- translationTest.py
- vibrationTest.py
```
including: 
- PEP-8 reformatting
- PEP-8 renaming of internal variables
- Futurize files
- Replace relative imports (`__init__.py`)
- remove unused `math` module

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -36,75 +36,80 @@
-            [ -0.7609, -1.2051, -1.6516],
-            [ -0.3125, -2.1468, -0.2292],
-            [ -1.8039, -1.2051, -0.2293],
+            [0.0000,  0.0000,  0.0000],
```
I think the original intention was to have all numbers align nicely in these lists. Since we're not doing that anymore with the repetitive spaces, could you replace any multiple spaces `_ _` with a single one `_`?
(I'm using the `_` sign for a space for visibility in this comment)

xiaoruiDong replied:
Thank Alon for changing it.

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -36,75 +36,80 @@
         ])
         self.number = numpy.array([6, 1, 1, 1, 6, 1, 1, 1])
-        self.mass = numpy.array([12, 1.007825, 1.007825, 1.007825, 12, 1.007825, 1.007825, 1.007825])
+        self.mass = numpy.array([12, 1.007825, 1.007825, 1.007825,
```
If this fits nicely into one line within the recommended max width, I prefer to keep the original version (less lines)

xiaoruiDong replied:
I am using Atom instead of PyCharm and the AutoPEP8 in Atom uses line width = 79 characters. I think it might be the reason my package suggested an extra line here. But it shouldn't be a big deal. I will keep your modification.

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -226,7 +233,8 @@ def test_getTotalMass(self):
         """
-        self.assertAlmostEqual(self.conformer.getTotalMass()*constants.Na*1000., numpy.sum(self.mass), 6)
+        self.assertAlmostEqual(self.conformer.getTotalMass() *
+                               constants.Na*1000., numpy.sum(self.mass), 6)
```
I think additional spaces around `*` are required

mliu49 replied:
I think it would also be preferable to split after the comma instead of in the middle of the expression.

xiaoruiDong replied:
Modified.

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -242,15 +250,15 @@ def test_getMomentOfInertiaTensor(self):
-        self.assertAlmostEqual(I[2,0]*constants.Na*1e23, -3.46416, 4)
-        self.assertAlmostEqual(I[2,1]*constants.Na*1e23, -5.48630, 4)
-        self.assertAlmostEqual(I[2,2]*constants.Na*1e23, 22.84296, 4)
+        self.assertAlmostEqual(I[0, 0]*constants.Na*1e23, 20.65968, 4)
```
additional spaces around the `*` operator

xiaoruiDong replied:
Modified

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -99,71 +100,72 @@ def test_getLevelDegeneracy(self):
-        Tlist = numpy.array([300,500,1000,1500,2000])
-        Hexplist = numpy.array([0.997705, 0.998624, 0.999312, 0.999541, 0.999656]) * constants.R * Tlist
+        Tlist = numpy.array([300, 500, 1000, 1500, 2000])
+        Hexplist = numpy.array([0.997705, 0.998624, 0.999312, 0.999541,
```
I think it's OK to leave that in just one line

alongd commented on rmgpy/statmech/torsionTest.py:
```diff
@@ -109,158 +111,161 @@ def test_getPotential_fourier(self):
-        Qexplist = numpy.sqrt(8*numpy.pi**3*constants.kB*Tlist*self.freemode.inertia.value_si)/(self.symmetry*constants.h)
-        for T, Qexp in zip(Tlist,Qexplist):
+        Tlist = numpy.array([300, 500, 1000, 1500, 2000])
+        Qexplist = numpy.sqrt(8*numpy.pi**3*constants.kB*Tlist *
```
spaces around `*` and `**` operators

alongd commented on rmgpy/statmech/vibrationTest.py:
```diff
@@ -41,90 +41,92 @@
-        Tlist = numpy.array([300,500,1000,1500,2000])
-        Hexplist = numpy.array([0.280395, 0.637310, 1.30209, 1.70542, 1.96142]) * constants.R * Tlist
+        Tlist = numpy.array([300, 500, 1000, 1500, 2000])
+        Hexplist = numpy.array([0.280395, 0.637310, 1.30209, 1.70542,
```
one line?

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -110,20 +110,20 @@ def test_getPartitionFunction_ethylene(self):
+        q_exp_list = numpy.array([4.05311e+09, 4.19728e+10, 2.82309e+12, 7.51135e+13, 1.16538e+15])
+        for T, q_exp in zip(Tlist, q_exp_list):
+            q_act = self.ethylene.getPartitionFunction(T)
+            self.assertAlmostEqual(q_exp, q_act, delta=1e-4*q_exp)
```
spaces around `*`

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -192,42 +192,43 @@ def test_getEnthalpy_oxygen(self):
         for n in range(10, len(Elist)):
             self.assertTrue(0.8 < numpy.sum(
-                densStates[0:n+1]) / sumStates[n] < 1.25, '{0} != {1}'.format(numpy.sum(densStates[0:n+1]), sumStates[n]))
+                dens_states[0:n+1]) / sum_states[n] < 1.25, '{0} != {1}'.format(numpy.sum(dens_states[0:n+1]), sum_states[n]))
```
`[0: n + 1]` with spaces (twice)

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -192,42 +192,43 @@ def test_getEnthalpy_oxygen(self):
-        self.assertAlmostEqual(Qexp, Qact, delta=1e-1*Qexp)
+        q_act = numpy.sum(dens_states * numpy.exp(-Elist / constants.R / T))
+        q_exp = self.oxygen.getPartitionFunction(T)
+        self.assertAlmostEqual(q_exp, q_act, delta=1e-1*q_exp)
```
spaces around `*` (and line 226)

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -108,10 +108,10 @@ def test_getPartitionFunction_classical(self):
-            Qact = self.mode.getPartitionFunction(T)
-            self.assertAlmostEqual(Qexp, Qact, delta=1e-4*Qexp)
+        q_exp_list = numpy.array([72.6691, 121.115, 242.230, 363.346, 484.461])
+        for T, q_exp in zip(Tlist, q_exp_list):
```
maybe `temperature` instead of `T`? (we don't want to use `t` either)
Same for the lines below

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -108,10 +108,10 @@ def test_getPartitionFunction_classical(self):
+        q_exp_list = numpy.array([72.6691, 121.115, 242.230, 363.346, 484.461])
+        for T, q_exp in zip(Tlist, q_exp_list):
+            q_act = self.mode.getPartitionFunction(T)
+            self.assertAlmostEqual(q_exp, q_act, delta=1e-4*q_exp)
```
spaces around `*`, same for the lines below

alongd commented on rmgpy/statmech/torsionTest.py:
```diff
@@ -117,11 +117,11 @@ def test_getPartitionFunction_free(self):
-        for T, Qexp in zip(Tlist, Qexplist):
-            Qact = self.freemode.getPartitionFunction(T)
-            self.assertAlmostEqual(Qexp, Qact, delta=1e-4*Qexp)
+        q_exp_list = numpy.sqrt(8*numpy.pi**3*constants.kB*Tlist *
```
spaces around operators, same for the lines below

xiaoruiDong replied:
Modified, but keep `numpy.pi**3`, I think it is better to distinguish 'power'.

alongd replied:
I added spaces around the power operator in my PRs, let decide on a consistent approach (whatever everyone agrees on)

alongd commented on rmgpy/statmech/torsionTest.py:
```diff
@@ -117,11 +117,11 @@ def test_getPartitionFunction_free(self):
-            self.assertAlmostEqual(Qexp, Qact, delta=1e-4*Qexp)
+        q_exp_list = numpy.sqrt(8*numpy.pi**3*constants.kB*Tlist *
+                                self.freemode.inertia.value_si)/(self.symmetry*constants.h)
+        for T, q_exp in zip(Tlist, q_exp_list):
```
`temperature`? same for the below

xiaoruiDong replied:
Modified.

alongd commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -32,6 +32,8 @@
 This script contains unit tests of the :mod:`rmgpy.statmech.conformer` module.
 """
 
+from __future__ import division
```
Do we need to keep this statement? Wasn't it added just for `futerize` to function as we intended? @mliu49

amarkpayne replied:
Correct, we do not need this statement and it should be removed.

mliu49 replied:
I've been keeping the division imports so that we can confirm that we won't break anything by using float division while we're testing in Python 2.

amarkpayne replied:
You said though that this was fine to remove if we have inspected all of the division cases and also ran the tests to confirm, right?

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -32,6 +32,10 @@
 """
 
+from __future__ import division
+from future import standard_library
```
I'm unfamiliar with this, could you teach me what it does?

mliu49 replied:
There are certain functions which were renamed in Python 3. This (and the line below) makes it so that you can use the Python 3 names and get the correct behavior.

In this specific case, it's because pickle/cPickle were combined into pickle in Python 3, which automatically decides whether to use the Python or C implementation.

alongd commented:
@xiaoruiDong, I added a commit with many changes suggested by PyCharm's code inspection (that I manually accepted). They also address (only) some of the comments I made on the original commits of this PR.

alongd commented:
Tests fail due to future imports. Perhaps we need to add future to the requirements?
```
======================================================================
ERROR: Failure: ImportError (No module named future)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/rmg_env/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/travis/miniconda/envs/rmg_env/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/travis/miniconda/envs/rmg_env/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/travis/build/ReactionMechanismGenerator/RMG-Py/rmgpy/statmech/rotationTest.py", line 36, in <module>
    from future import standard_library
ImportError: No module named future
```

mliu49 commented:
We should remove that import since we're not trying to be compatible with 2 and 3.

mliu49 commented on rmgpy/statmech/torsionTest.py:
```diff
@@ -397,8 +406,9 @@ def test_getSumOfStates_quantum_fourier(self):
-            self.assertTrue(0.8 < numpy.sum(densStates[0:n]) / sumStates[n-1] < 1.25, '{0} != {1}'.format(numpy.sum(densStates[0:n]), sumStates[n]))
-        
+            self.assertTrue(0.8 < numpy.sum(
+                densStates[0:n]) / sumStates[n-1] < 1.25, '{0} != {1}'.format(numpy.sum(densStates[0:n]), sumStates[n]))
```
This should be split after the comma rather than in the middle of the expression.

xiaoruiDong replied:
Changed.

mliu49 commented on rmgpy/statmech/vibrationTest.py:
```diff
@@ -177,17 +180,19 @@ def test_getSumOfStates_quantum(self):
             else:
-                self.assertTrue(0.8 < numpy.sum(densStates[0:n]) / sumStates[n-1] < 1.25, '{0} != {1}'.format(numpy.sum(densStates[0:n]), sumStates[n]))
+                self.assertTrue(0.8 < numpy.sum(
+                    densStates[0:n]) / sumStates[n-1] < 1.25, '{0} != {1}'.format(numpy.sum(densStates[0:n]), sumStates[n]))
```
Split after comma

xiaoruiDong replied:
Changed.

mliu49 commented on rmgpy/statmech/schrodingerTest.py:
```diff
@@ -63,54 +63,54 @@ def test_getPartitionFunction(self):
 #        sumStates = getSumOfStates(Elist, self.energy, self.degeneracy, self.n0)
 #        for n in range(1, len(Elist)):
-#            self.assertAlmostEqual(numpy.sum(densStates[0:n+1]) / sumStates[n], 1.0, 3)
+#            self.assertAlmostEqual(numpy.sum(dens_states[0:n+1]) / sumStates[n], 1.0, 3)
```
If you're renaming dens_states, you might as well rename sum_states, even though this test is commented out.

xiaoruiDong replied:
Changed.

xiaoruiDong commented:
Thank everyone @alongd @mliu49 @amarkpayne  for the comments, I will update as requested.

alongd commented on rmgpy/statmech/__init__.py:
```diff
@@ -28,8 +28,12 @@
+from rmgpy.statmech.rotation import Rotation, LinearRotor, NonlinearRotor, KRotor, SphericalTopRotor
+from rmgpy.statmech.vibration import Vibration, HarmonicOscillator
+from rmgpy.statmech.torsion import Torsion, HinderedRotor
+from rmgpy.statmech.conformer import Conformer
```
Could you reorder the imports alphabetically? i.e., start with the `.conformer` one, etc.
Also, take a look at other files in this PR?

xiaoruiDong replied:
I changed imports accordingly. I split the imports into 3 sections: (1) future (2) numpy, os, unittest,etc. (3)rmg related imports. I change the sequence of numpy and unittest. If you think it is inappropriate, I can change it back.

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -53,191 +55,191 @@ def setUp(self):
+        dens_states = self.mode.getDensityOfStates(e_list)
+        sum_states = self.mode.getSumOfStates(e_list)
+        for n in range(1, len(e_list)):
+            self.assertAlmostEqual(np.sum(dens_states[0:n]) / sum_states[n], 1.0, 3)
```
Could you replace `[0:n]` with `[0: n]`?

mliu49 replied:
Since this is slicing, the spacing on both sides of the `:` should be equivalent. I think we prefer no spaces.

xiaoruiDong replied:
I am using tight style now, like `[0:n + 1]`

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -250,129 +252,131 @@ def test_repr(self):
         """
-        import cPickle
-        mode = cPickle.loads(cPickle.dumps(self.mode,-1))
+        import pickle
```
Should this import be relocated to the top of the file, or are we testing importing it as well?

mliu49 replied:
All of the pickle tests import in the test itself. I'm not sure if there's a particular reason for that, but I think we can leave it for now. We should change it in all test modules if we do decide to change it though.

alongd commented on rmgpy/statmech/rotationTest.py:
```diff
@@ -382,33 +386,34 @@ def test_repr(self):
         """
-        import cPickle
-        mode = cPickle.loads(cPickle.dumps(self.mode,-1))
+        import pickle
```
same question

alongd commented on rmgpy/statmech/torsionTest.py:
```diff
@@ -32,426 +32,433 @@
+        sum_states = self.mode.getSumOfStates(e_list)
+        dens_states = self.mode.getDensityOfStates(e_list)
+        for n in range(10, len(e_list)):
+            self.assertTrue(0.8 < np.sum(dens_states[0: n]) / sum_states[n-1] < 1.25,
```
spaces around `-`

mliu49 replied:
I agree with spaces around `-`, but we should remove the space after `:` in this instance.

xiaoruiDong replied:
now [0:n - 1]

alongd commented on rmgpy/statmech/vibrationTest.py:
```diff
@@ -32,183 +32,188 @@
+        dens_states = self.mode.getDensityOfStates(e_list)
+        for n in range(1, len(e_list)):
+            if sum_states[n - 1] == 0:
+                self.assertTrue(np.sum(dens_states[0: n]) == 0,
```
This test seems strange. Perhaps lets do `assertFalse(expression)` instead of `assertTrue(expression == 0)`.
Also, the message (`'{0} != {1}'`) is trivial, we can get rid of it.

mliu49 replied:
I think `assertFalse` might be more general than necessary. It should be changed to `assertEqual` though, which has that as the default failure message.

xiaoruiDong replied:
Now using `assertEqual` without extra message. I also remove trivial message for some `assertAlmostEqual`

xiaoruiDong commented:
Thanks, @alongd and @mliu49 for revision!

mliu49 commented on rmgpy/statmech/conformerTest.py:
```diff
@@ -34,12 +34,12 @@
 
-import unittest
 import numpy as np
+import unittest
```
`unittest` is actually part of the standard library, so it should go in the first section, before numpy.

mliu49 commented:
I added one minor comment. I think this is good to go otherwise. You can go ahead and squash everything into one commit, since it's all changes within one module.

xiaoruiDong commented:
Thanks for the information, Max! I have squashed all the commits into one and reworded the commit message. I also rebased my branch with the current origin/py3 branch. Let me know if it is the right way and if there is something else I need to do. Thanks!

mliu49 commented:
Sorry, two more things.
1. Could you add an empty line between `import unittest` and `import numpy` if all relevant files?
2. Could you clean up the commit message so it's not just the messages from all the commits together?

xiaoruiDong commented:
Thanks @mliu49. Changes have been made accordingly.
